### ISSUE TITLE
Synchronous storage volume API endpoints may result in blocking server operations

### ISSUE THREAD

**Issue by mas-who** [canonical/lxd#12637]
# Required information
 * Distribution: Ubuntu snap
 * Distribution version: 5.19

# Issue description

The api endpoints for updating a volume `PUT /1.0/storage-pools/{poolName}/volumes/{type}/{volumeName}` and updating a volume snapshot `PUT /1.0/storage-pools/{poolName}/volumes/{type}/{volumeName}/snapshots/{snapshotName}` are synchronous at the moment. Making a request to either endpoint results in the below successful response:
```
{
  type: "sync",
  status: "Success",
  status_code: 200,
  operation: "",
  error_code: 0,
  error: "",
  metadata: {},
}
```
This seems to be inconsistent with other API endpoints which typically results in asynchronous operations. This is concerning since it may result in long and blocking operations on the server. Suggestion is to make these endpoints asynchronous.

**Comment by tomponline** [comment]
@MusicDin this has been reported from our LXD UI team so would be good to take a look and evaluate.

**Comment by tomponline** [comment]
@MusicDin please can you look at this one as your next bug, as if we need to change the API behaviour to make it consistent then we need to do so before LXD 5.21 LTS is released.

**Comment by tomponline** [comment]
@MusicDin if we made these endpoints async would this affect your devlxd APIs? @mionaalex this might be something we want to do before 6 LTS as its an api break.

**Comment by MusicDin** [comment]
> If we made these endpoints async would this affect your devlxd APIs?

Yes, in the same way the main LXD API is affected - the operation needs to be returned and potentially waited for on the client side. 

Should I (re)open a PR for this?

**Comment by tomponline** [comment]
@MusicDin did you previously envision supporting both sync and async modes? 

Also would the updated SDK just return a nil-operation when talking with an older LXD without the API extension?

**Comment by tomponline** [comment]
@MusicDin presumably it would make sense to decide upon this one now before we land the devlxd routes?

**Comment by MusicDin** [comment]
Hmm. I'm all for updating those endpoints, I'm just a bit concerned that it would even further postpone the CSI. How about I do the first PR addressing the routes that are included in initial devLXD volume management? And afterwards the remaining routes?

**Comment by MusicDin** [comment]
> did you previously envision supporting both sync and async modes?
> Also would the updated SDK just return a nil-operation when talking with an older LXD without the API extension?

I see the situation like this:

|            |  Old client | New Client
| ---------- | ----------- | ---
| Old server | ok          | Nil op. / "empty" operation? 
| New server | Incompatible - Client is not aware the operation is now being returned           | ok

The main breaking change is for old client. When *old* client communicates with the *new* server, it is not aware of the operation being received. Therefore, a successful response would not necessary indicate a successful action.

The less concerning is new client against old server. In such case we could return nil operation, or maybe even better some sort of *empty* operation to avoid returning nil on success?

**Comment by tomponline** [comment]
Agree. A faux operation would be nice indeed.

We don't need to support forward compatibility with older clients on newer severs.

**Comment by tomponline** [comment]
> Hmm. I'm all for updating those endpoints, I'm just a bit concerned that it would even further postpone the CSI. How about I do the first PR addressing the routes that are included in initial devLXD volume management? And afterwards the remaining routes?

this seems like a reasonable approach to me, agree @skozina @mionaalex ?

**Comment by MusicDin** [comment]
The list of affected endpoints (those that do not return operation):
```
POST   /storage-pools/<pool>/volumes/<volType>                            # Create storage volume
PUT    /storage-pools/<pool>/volumes/<volType>/<vol>                      # Update storage volume
PATCH  /storage-pools/<pool>/volumes/<volType>/<vol>                      # Update storage volume
DELETE /storage-pools/<pool>/volumes/<volType>/<vol>                      # Delete storage volume
PUT    /storage-pools/<pool>/volumes/<volType>/<vol>/snapshots/<snapshot> # Update storage volume snapshot
POST   /storage-pools/<pool>/buckets                                      # Create storage bucket
PUT    /storage-pools/<pool>/buckets/<bucket>                             # Update storage bucket
DELETE /storage-pools/<pool>/buckets/<bucket>                             # Delete storage bucket
```

EDIT:
As Mark pointed out in https://github.com/canonical/lxd/issues/12637#issuecomment-3468637698, the following endpoints will be affected as well. 
```
PUT /1.0/profiles/{name}
PATCH /1.0/profiles/{name}
```

**Comment by tomponline** [comment]
@MusicDin chatted with @escabo and we've decided to hold off on this for now, but land the changes before 6.6.

**Comment by markylaing** [comment]
We should review any other endpoints that should return an operation but don't. The only ones that come to mind for me are:
- `PUT /1.0/profiles/{name}`
- `PATCH /1.0/profiles/{name}`

**Comment by MusicDin** [comment]
@tomponline @markylaing Should we also change `POST / DELETE` for profiles to return an operation. This would be consistent with other endpoints (instance, volumes, etc.)

**Comment by tomponline** [comment]
Can those operations take a long time (with any options like force if applicable?) or do they just create or delete a dB entry?

**Comment by MusicDin** [comment]
> Can those operations take a long time (with any options like force if applicable?) or do they just create or delete a dB entry?

No. Currently they do not. They just mangle the db entries.

**Comment by tomponline** [comment]
Let's leave as is then thanks

**Comment by MusicDin** [comment]
The final list of changed endpoints (storage buckets were omitted):
```
POST   /storage-pools/<pool>/volumes/<volType>                            # Create storage volume
PUT    /storage-pools/<pool>/volumes/<volType>/<vol>                      # Update storage volume
PATCH  /storage-pools/<pool>/volumes/<volType>/<vol>                      # Patch storage volume
POST   /storage-pools/<pool>/volumes/<type>/<vol>                         # Rename storage volume
DELETE /storage-pools/<pool>/volumes/<volType>/<vol>                      # Delete storage volume

PUT    /storage-pools/<pool>/volumes/<volType>/<vol>/snapshots/<snapshot> # Update storage volume snapshot

PUT    /1.0/profiles/<profile>                                            # Update profile
PATCH  /1.0/profiles/<profile>                                            # Patch profile
```

API extension: [`storage-and-profile-operations`](https://documentation.ubuntu.com/lxd/latest/api-extensions/#storage-and-profile-operations)

@edlerd This is now implemented and should be available in `latest/edge`.

**Comment by tomponline** [comment]
@MusicDin can we close this?
